### PR TITLE
Start advanced numpy lesson

### PR DESCRIPTION
**Pull request by wmvanvliet**
A new lesson covering advanced NumPy stuff that the basic tutorial won't teach you.

### CODE REVIEW

**Comment by wmvanvliet** [comment]
Still a work in progress

File: content/numpy-advanced.rst [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+   The function can be implemented like this::
+
+     def ravel(row, col, n_rows, n_cols):
+       return row * n_cols + col
+
+Strides
+-------
+
+As seen in the exercise, to get to the next row, we have to skip over
+``n_cols`` indices. To get to the next column, we can just add 1. To generalize
```
Comment (rkdarst) [inline content/numpy-advanced.rst:137]: Not right here (maybe in the basic numpy lesson), but do you have any insight for how to develop intuition for what is a "row" and what is a "column"?

File: content/numpy-advanced.rst [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+this code to work with an arbitrary number of dimensions, NumPy has the concept
+of "strides"::
+
+  np.zeros((4, 8)).strides           # (64, 8)
+  np.zeros((4, 5, 6, 7, 8)).strides  # (13440, 2688, 448, 64, 8)
+
+The ``.strides`` attribute contains for each dimension, the number of bytes we
```
Comment (rkdarst) [inline content/numpy-advanced.rst:144]:
```suggestion
The ``.strides`` attribute contains for each dimension, the number of *bytes* (not array indexes) we
```

File: content/numpy-advanced.rst [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+  a = rng.rand(10_000, 20_000)
+  print(f'Matrix `a` takes up {a.nbytes / 10**6} MB')
+
+Let's time the :func:`numpy.transpose` function::
+
+  %%timeit
+  b = a.transpose()
+
+It takes mere nanoseconds to transpose 1600 MB of data! NumPy avoided copying
+any data by modifying the ``.strides``::
```
Comment (rkdarst) [inline content/numpy-advanced.rst:169]:
```suggestion
any data by *only* modifying the ``.strides`` of the existing array in-place::
```

File: content/numpy-advanced.rst [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+  
+  print('\nTransposed and then reshaped:')
+  print(a.T.reshape(2, 3))
+
+
+Copy versus view
+----------------
+
+Whenever NumPy constructs a new array by modifying the ``.strides`` instead of
+copying data, we way it created a "view". This also happens when we select only
```
Comment (rkdarst) [inline content/numpy-advanced.rst:276]: So, by this point (now or earlier), should we mention how array object = metadata (in the object) + buffer (is it ro or rw), and views use the same buffer with different offset(?) and strides?  At least for me, that's what I am missing when I get to this point.

**Review by rkdarst — COMMENTED** [review]
I didn't mean to leave comments, but I thought of things and it was easier to write than remember.  Feel free to ignore.

File: content/numpy-advanced.rst (hunk 1 above)
Comment (wmvanvliet) [reply content/numpy-advanced.rst:137]: These things are called columns, they are tall vertical things:
column

**Review by johanhellsvik — COMMENTED** [review]
The contents and the presentation of the lesson are good.
The numbering of exercises needs to be revised. The first exercise is labeled Exercise 2.

**Comment by wmvanvliet** [comment]
Perhaps rather, exercise 1 is missing :)

**Review by johanhellsvik — APPROVED** [review]

**Pull request merged**